File: python_examples/Adjust Settings Example.py
import sys

import pythoncom
import clr

pythoncom.CoInitialize()

# EA Engine installation path
installPath = "C:\\Program Files\\Bruel and Kjaer\\EA Engine"
# Add the EA Engine .NET assembly (EA Engine executable filename - including full path)
clr.AddReference("C:\\Program Files\\Bruel and Kjaer\\EA Engine\\EA_Engine.exe")
# Import the assembly namespace
from EA_Engine import *

# Declare and instanciate a new Engine object (the installation path is mandatory)
engine = Engine(installPath)

# Get version of the EA Engine installed
print(engine.GetVersion())

# Get EA Engine settings
engineSettings = engine.GetEngineSettings()
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()
else:
    print("EA Engine settings:")
    print("Data Folder: ", engineSettings.DataFolder)
    print("Settling Time (s): ", engineSettings.SettlingTime)
    print("Integrate/Differentiate: ", engineSettings.IntegrateDifferentiate)
    print("Calibration Threshold: ", engineSettings.CalibrationThreshold)
    print("\n")

# Set EA Engine settings (for example, set settling time to 1 s and ask to differenciate once)
engineSettings = engine.SetEngineSettings(engineSettings.DataFolder, 1, 1, engineSettings.CalibrationThreshold)
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# Detect ASIO devices
devices = engine.DetectDevices()
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# List detected devices
print("Detected devices:")
for device in devices:
    print("Index: ", device.Index)
    print("Name: ", device.Name)
    print("ASIO Driver Name: ", device.ASIODriverName)
    print("Selected: ", device.Selected)
    print("\n")

# Select the first device (index starts at 1) and keep existing Input Channels.xml and Output Channels.xml files
device = engine.SelectDevice(1, False)
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# Get detected devices
devices = engine.GetDetectedDevices()
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# List (again) detected devices - the selected device will have now its Selected property set to true
print("Detected devices once one has been selected:")
for device in devices:
    print("Index: ", device.Index)
    print("Name: ", device.Name)
    print("ASIO Driver Name: ", device.ASIODriverName)
    print("Selected: ", device.Selected)
    print("\n")

# Get selected device capabilities
deviceCapabilities = engine.GetSelectedDeviceCapabilities()
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# List the selected device capabilities
print("Selected device capabilities:")
print("Sample Rate (Hz): {0:.0f}".format(deviceCapabilities.SampleRate))
print("Buffer Size: ", deviceCapabilities.BufferSize)
print("Input Latency (samples): ", deviceCapabilities.InputLatency)
print("Input Latency (ms): {0:.2f}".format(deviceCapabilities.InputLatencyTime))
print("Output Latency (samples): ", deviceCapabilities.OutputLatency)
print("Output Latency (ms): {0:.2f}".format(deviceCapabilities.OutputLatencyTime))
print("\n")

# Get input channels
inputChannels = engine.GetInputChannels(False)
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# Set first input channel (first way)
inputChannels[0] = engine.SetInputChannel(1, "4944-A", True, "4227-A", 0.84, "mV/Pa", 2e-5)
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# Set first input channel (second way)
inputChannels[1].Name = "Input Channel 2"
inputChannels[1].IsActive = False
inputChannels[1].ReferenceChannelName = "None"
inputChannels[1].Sensitivity = 1000
inputChannels[1].SensitivityUnit = "mV/V"
inputChannels[1].dBRef = 1
engine.SetInputChannels(inputChannels)
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# Get output channels
outputChannels = engine.GetOutputChannels(False)
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# Set first output channel (first way)
outputChannels[0] = engine.SetOutputChannel(1, "4227-A", True, "None", 1, "V/V", 1, GeneratorSignalType.White)
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

# Set first output channel (second way)
outputChannels[1].Name = "Generator 2"
outputChannels[1].IsActive = False
outputChannels[1].ReferenceChannelName = "None"
outputChannels[1].Sensitivity = 500
outputChannels[1].SensitivityUnit = "mV/V"
outputChannels[1].dBRef = 1
outputChannels[1].SignalType = GeneratorSignalType.Sine
engine.SetOutputChannels(outputChannels)

# Set first output Random settings (used when its signal type is set to White or Pink)
randomSettings = engine.SetOutputChannelRandomSettings(1, 0.025, True, 80, 12000, 0, 5)
if engine.ErrorMessage != "":
    print(engine.ErrorMessage)
    sys.exit()

File: matlab_examples/AdjustSettingsExample.m
clc;
clear all;

% EA Engine installation path
installPath = 'C:\Program Files\Bruel and Kjaer\EA Engine';
% EA Engine executable filename (including full path)
assemblyName = 'C:\Program Files\Bruel and Kjaer\EA Engine\EA_Engine.exe';

% Add the EA Engine .NET assembly (please note that it cannot be removed from the current MATLAB session until exiting)
EA_Engine = NET.addAssembly(assemblyName);
% Import the assembly namespace
import EA_Engine.*;

% Declare and instanciate a new Engine object (the installation path is mandatory)
engine = Engine(installPath);

% Get version of the EA Engine installed
disp(engine.GetVersion());

% Get EA Engine settings
engineSettings = engine.GetEngineSettings();
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% List EA Engine settings
fprintf('EA Engine settings:\n');
fprintf('Data Folder : %s\n', string(engineSettings.DataFolder));
fprintf('Settling Time (s): %d\n', engineSettings.SettlingTime);
fprintf('Integrate/Differentiate: %d\n', engineSettings.IntegrateDifferentiate);
fprintf('Calibration Threshold: %d\n', engineSettings.CalibrationThreshold);
fprintf('\n');

% Set EA Engine settings (for example, set settling time to 0 s
engineSettings = engine.SetEngineSettings(engineSettings.DataFolder, 0, 0, engineSettings.CalibrationThreshold);
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% Detect ASIO devices
devices = engine.DetectDevices();
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% List detected devices
fprintf('Detected devices:\n');
for i = 1:length(devices)
   fprintf('Index: %d\n', devices(i).Index);
   fprintf('Name: %s\n', string(devices(i).Name));
   fprintf('ASIO Driver Name: %s\n', string(devices(i).ASIODriverName));
   fprintf('Selected: %s\n', mat2str(devices(i).Selected));
   fprintf('\n');
end

% Select the first device (index starts at 1) and keep existing Input Channels.xml and Output Channels.xml files
device = engine.SelectDevice(1, false);
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% Get detected devices
devices = engine.GetDetectedDevices();
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% List (again) detected devices - the selected device will have now its Selected property set to true
fprintf('Detected devices once one has been selected:\n');
for i = 1:length(devices)
   fprintf('Index: %d\n', devices(i).Index);
   fprintf('Name: %s\n', string(devices(i).Name));
   fprintf('ASIO Driver Name: %s\n', string(devices(i).ASIODriverName));
   fprintf('Selected: %s\n', mat2str(devices(i).Selected));
   fprintf('\n');
end

% Get selected device capabilities
deviceCapabilities = engine.GetSelectedDeviceCapabilities();
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% List the selected device capabilities
fprintf('Selected device capabilities:\n');
fprintf('Sample Rate (Hz): %d\n', deviceCapabilities.SampleRate);
fprintf('Buffer Size: %d\n', deviceCapabilities.BufferSize);
fprintf('Input Latency (samples): %d\n', deviceCapabilities.InputLatency);
fprintf('Input Latency (ms): %.2f\n', deviceCapabilities.InputLatencyTime);
fprintf('Output Latency (samples): %d\n', deviceCapabilities.OutputLatency);
fprintf('Output Latency (ms): %.2f\n', deviceCapabilities.OutputLatencyTime);
fprintf('\n');

% Get input channels
inputChannels = engine.GetInputChannels(false);
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% Set first input channel (first way)
inputChannels(1) = engine.SetInputChannel(1, '4944-A', true, '4227-A', 0.84, 'mV/Pa', 2e-5); 
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% Set first input channel (second way)
inputChannels(2).Name = 'Input Channel 2';
inputChannels(2).IsActive = false;
inputChannels(2).ReferenceChannelName = 'None';
inputChannels(2).Sensitivity = 1000;
inputChannels(2).SensitivityUnit = 'mV/V';
inputChannels(2).dBRef = 1;
engine.SetInputChannels(inputChannels);
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% Get output channels
outputChannels = engine.GetOutputChannels(false);
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% Set first output channel (first way)
outputChannels(1) = engine.SetOutputChannel(1, '4227-A', true, 'None', 1, 'V/V', 1, GeneratorSignalType.White); 
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% Set first output channel (second way)
outputChannels(2).Name = 'Generator 2';
outputChannels(2).IsActive = false;
outputChannels(2).ReferenceChannelName = 'None';
outputChannels(2).Sensitivity = 500;
outputChannels(2).SensitivityUnit = 'mV/V';
outputChannels(2).dBRef = 1;
outputChannels(2).SignalType = GeneratorSignalType.Sine;
engine.SetOutputChannels(outputChannels);
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end

% Set first output Random settings (used when its signal type is set to White or Pink)
randomSettings = engine.SetOutputChannelRandomSettings(1, 0.025, true, 80, 12000, 0, 5);
if (engine.ErrorMessage ~= "")
    disp(engine.ErrorMessage);
    return;
end
